### PR TITLE
Improve sensitivity estimator

### PR DESCRIPTION
**Pull request by mackaiver**
Hello,
I'm reopening my old PR for the sensitivity calculation

I'm using scipy's `newton` method  here to find the root of 

```
significance_on_off(on, off, alpha, method='lima') - self.sigma
```

This simplifies the method which estimated the number of excess counts needed to reach the target significance. The resulting excess numbers are within one or two decimals of the old method. The first test case fails however. I'm curious where the numbers in the unit tests come from. Is it some objective truth calculated by another program?

### CODE REVIEW

**Comment by cdeil** [comment]
@mackaiver - Thanks!

I fixed the test fail in `gammapy/scripts/tests/test_cta_sensitivity.py` in 05c1e302 in master, because I want to get a clean CI build. @bkhelifi Could you please look at the new values in the assert and comment if they are OK or not. I don't know why they changed.

I wanted to be careful and not touch lines that would conflict with this PR, but apparently wasn't and now there's a conflict in `gammapy/scripts/tests/test_cta_sensitivity.py`. @mackaiver - Can you please rebase?

In addition, my suggestion here would be to change the implementation to always add things that are computed (like excess) to a single results table. It's easier to use, and also the tests can be combined if one just has to do "run" and then gets a result table with all information, no?
@mackaiver @bkhelifi - Thoughts on this point?

File: gammapy/scripts/cta_sensitivity.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 from gammapy.spectrum.models import PowerLaw
 from gammapy.spectrum.utils import CountsPredictor
 from gammapy.scripts import CTAPerf
+from scipy.optimize import newton
```
Comment (cdeil) [inline gammapy/scripts/cta_sensitivity.py]: In Gammapy we still have `scipy` as an optional dependency, and an `import scipy` at the top of this file will fail our CI build where we collect / run tests with just the core dependencies and no scipy installed.
Making `scipy` and `yaml` core dependencies is something we should discuss, but for now, @mackaiver - Please move the import into the method where you use it.

File: gammapy/scripts/cta_sensitivity.py [hunk 2]
```diff
@@ -110,47 +111,29 @@ def get_excess(self, bkg_counts):
...
-            coarse_sigma = significance_on_off(n_on=coarse_on, n_off=coarse_off, alpha=self.alpha, method='lima')
-            idx = np.abs(coarse_sigma - self.sigma).argmin()
-
-            start = coarse_excess[max(idx - 1, 0)]
-            stop = coarse_excess[min(idx + 1, len(coarse_sigma) - 1)]
-            if start == stop:
-                log.warning('LOGICAL ERROR> Impossible to find a number of gamma!')
-                excess[icount] = -1
+        excess = np.zeros_like(bkg_counts)
+        for energy_bin, bg_count in enumerate(bkg_counts):
```
Comment (cdeil) [inline gammapy/scripts/cta_sensitivity.py]: Suggest we change "bg" -> "bkg" for uniformity in naming here.
(and if needed in follow-up PRs in other parts of the Gammapy code as well, but here please keep changes to this one file / class)

File: gammapy/scripts/cta_sensitivity.py (hunk 1 above)
Comment (bkhelifi) [reply gammapy/scripts/cta_sensitivity.py]: done

File: gammapy/scripts/cta_sensitivity.py (hunk 2 above)
Comment (bkhelifi) [reply gammapy/scripts/cta_sensitivity.py]: done

**Comment by cdeil** [comment]
@bkhelifi - In Gammapy master we have this fail for `test_cta_irf` because it's accessing an example datafile where the `SENSITIVITY` HDU doesn't exist!?
https://travis-ci.org/gammapy/gammapy/jobs/275788705#L2488

Can you make it better (here or in a separate PR) today? If no, I'll disable that test today, so that we have a clean CI build for Gammapy and this doesn't show up as an unrelated fail in all other pull requests.

**Comment by bkhelifi** [comment]
I have inserted the modifications and tested, via the command line instructions. All are now in GitHub master! Meaning that the PR is accepted and finished

**Comment by cdeil** [comment]
@bkhelifi - The commits from this PR or your follow-up commits aren't on Github yet:
https://github.com/gammapy/gammapy/commits/master

Currently the last commit is 2f916bfc6a3a5427f93d63fa3b18803a6bddfd31

Probably you still have to push the new version?
Please leave a comment here with a link to the merge commit where you added the changes from this PR, so that there's a proper record of what happened.

**Comment by bkhelifi** [comment]
Really! ok... it should be done now.
How can one check that a pull is effectivelly done?

**Comment by cdeil** [comment]
You can view the commits in the `master` branch here:

https://github.com/gammapy/gammapy/commits/master

You can view the commits in this PR from @mackaiver here:

https://github.com/gammapy/gammapy/pull/1123/commits

The last commit in master is 2f916bfc6a3a5427f93d63fa3b18803a6bddfd31 , the three commits from this PR (e.g. c2ffcd4ea0abd8985e57b90e6c360a81341c55de ) aren't in master yet.

Unless you're very experienced with git, the best way is to never push to `master` directly, but instead to always go via pull requests, and only after reviewing the list of commits and the diff for a pull request to push the "merge" button. This makes sure there can be no git mistakes in the `master` branch. Although I have to admit, in this case it wasn't possible, because @mackaiver didn't rebase this PR, so merging was blocked, so it wasn't an easy case for your first PR review for Gammapy. Let's have a short Skype call to sort this out. I'm on Skype now.

**Comment by bkhelifi** [comment]
I have just followed strictly the 'command line instructions' for the PR. I am not experienced, so I am trying to follow the documentation as close as possible.

But, I see an apparent contradiction:
[11:35am] /home/bruno/MesProgrammes/gammapy> git branch
  mackaiver-bkhelifi_sensi
  maps
* master
  sensi
  spectrum
[11:49am] /home/bruno/MesProgrammes/gammapy> git status
Sur la branche master
Votre branche est en avance sur 'upstream/master' de 21 commits.
  (utilisez "git push" pour publier vos commits locaux)
[11:49am] /home/bruno/MesProgrammes/gammapy> git push bkhelifi master
Everything up-to-date

For me, the modified code is well on git!

**Comment by cdeil** [comment]
It's `git push <remote-alias> <local-branch-name>`.

What is the remote alias you have for the "main" repo on Github, the one at `gammapy/gammapy`?
Find out via
```
git remote -v
```
For me it's called `origin`, so the command to push the local `master` branch to the remote `master` branch in the main `gammapy/gammapy` repo on Github would be:
```
git push origin master
```
Try it. If there's a Problem I'll revert. (I'm going for lunch, will check back later)

**Comment by cdeil** [comment]
Ah, I see this in your console output:
```
Votre branche est en avance sur 'upstream/master' de 21 commits.
```
So for you, the alias for the main repo on Github is called "origin" and you should do this:
```
git push origin master
```

**Comment by cdeil** [comment]
Any why are there 21 commits!???

**Comment by bkhelifi** [comment]
no idea! It is from:
Modifications qui ne seront pas validées :
  (utilisez "git add <fichier>..." pour mettre à jour ce qui sera validé)
  (utilisez "git checkout -- <fichier>..." pour annuler les modifications dans la copie de travail)

	modifié:         astropy_helpers (nouveaux commits)

**Comment by mackaiver** [comment]
Why don't you just reopen this PR and add commits here? Then just merge it via the webinterface.

**Comment by bkhelifi** [comment]
I understand now! the main branch is called upstream for me (thanks for the command 'git remote -v').
It should be done now... hopefully

**Comment by cdeil** [comment]
@bkhelifi - I looked for 30 min with @adonath at the git history of Gammapy `master`, and it's extremely hard to understand, it's a tangled web of merge commits bringing in good and bad changes multiple times. There are a few changes that are really problematic one of them being the flux point upper limit computation that would change HGPS results which we are running at the moment.

So apologies, but I'm doing a hard reset from bec7a9640e12afb21139aafb2c21dfa13545c125 (your latest merge commit) to the state of Gammapy master as of an hour ago (2f916bfc6a3a5427f93d63fa3b18803a6bddfd31 ), before the commits from this PR and many other commits came in. Then @adonath will rebase this PR 1123 and merge it in, or open a new PR, i.e. integrate these changes.

For reference, here is the compare view between these two commits:
https://github.com/gammapy/gammapy/compare/2f916bfc6a3a5427f93d63fa3b18803a6bddfd31...bec7a9640e12afb21139aafb2c21dfa13545c125

Here's the current version history of Gammapy at https://github.com/gammapy/gammapy/network showing part of the complex tree of merge commits going backward:
screen shot 2017-09-15 at 14 13 03

@adonath or I will write some developer / maintainer documentation for Gammapy and explain them next Thursday, to make the git / Github workflow clear and pleasant for the future. Just very briefly: the Gammapy workflow is exactly the same as for Astropy, and one of the rules that leads to a simple version history is this:
> Never merge changes from astropy/master into your feature branch. If changes in the development version require changes to our code you can Rebase, but only if asked.

http://astropy.readthedocs.io/en/latest/development/workflow/development_workflow.html#astropy-guidelines-for-git

If this rule is followed, then the Gammapy commit history will remain simple:
* either completely linear (if someone experienced like @adonath  or I makes small commits directly in master)
* or simple branches / PRs of linear commits that get into `master` via a single merge commit (this is the usual procedure for almost all changes, and the only one except for small changes and people that have worked with git for a long time and fully understand it)

**Comment by cdeil** [comment]
For reference: Just now I
* De-activated `master` branch protection at https://github.com/gammapy/gammapy/settings/branches/master
* `git reset --hard 2f916bfc6a3a5427f93d63fa3b18803a6bddfd31`
* Re-activated `master` branch protection

@adonath - Please go ahead and integrate the changes from @mackaiver here. I would do it via a new branch / pull request and cherry-picking the commits, not directly in master. But as you like.

**Pull request closed without merging**